perses/rjmc/atom_mapping.py

```diff
@@ -0,0 +1,1475 @@
+                                                        bond_expr=self.bond_expr,
+                                                        matching_criterion=self.matching_criterion)
+                atom_mappings.update(atom_mappings_for_this_scaffold_map)
+
+        if not self.allow_ring_breaking:
+            # Filter the matches to remove any that allow ring breaking
+            _logger.debug(f'Fixing mappings to not create or break rings')
+            valid_atom_mappings = set()
+            for atom_mapping in atom_mappings:
+                try:
+                    atom_mapping.render_image('debug.png')
+                    atom_mapping.unmap_partially_mapped_cycles()
+                    valid_atom_mappings.add(atom_mapping)
+                except InvalidMappingException as e:
+                    # Atom mapping is no longer valid
+                    stop
+                    pass
+
+            atom_mappings = valid_atom_mappings
+
+        # TODO: Should we attempt to preserve chirality here for all atom mappings?
+        # Or is this just for biopolymer residues?
+
+        if len(atom_mappings) == 0:
+            _logger.warning('No maps found. Try relaxing match criteria or setting allow_ring_breaking to True')
+            return None
+
+        # Render set of AtomMapping to a list to return
+        return list(atom_mappings)
+
+    def get_best_mapping(self, old_mol, new_mol):
+        """Retrieve the best mapping between old and new molecules.
+
+        Parameters
+        ----------
+        old_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The initial molecule for the transformation.
+        new_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The final molecule for the transformation.
+
+        Returns
+        -------
+        atom_mapping : AtomMapping
+            Atom mapping with the best score
+
+        Examples
+        --------
+        Retrieve best-scoring mapping between ethane and ethanol
+
+        >>> from openff.toolkit.topology import Molecule
+        >>> ethane = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])([H:7])')
+        >>> ethanol = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])[O:7][H:8]')
+        >>> atom_mapping = atom_mapper.get_best_mapping(ethane, ethanol)
+
+        """
+        import time
+        initial_time = time.time()
+
+        import numpy as np
+        atom_mappings = self.get_all_mappings(old_mol, new_mol)
+        if (atom_mappings is None) or len(atom_mappings)==0:
+            return None
+
+        scores = np.array([ self.score_mapping(atom_mapping) for atom_mapping in atom_mappings ])
+        best_map_index = np.argmax(scores)
+
+        elapsed_time = time.time() - initial_time
+        _logger.debug(f'get_best_mapping took {elapsed_time:.3f} s')
+
+        return atom_mappings[best_map_index]
```

Comment:
Can we get multiple mappings with the same max score? What happens when that's the case?

Reply:
Great point! Added a comment to address this ambiguity.